network/src/sync/master.rs

```diff
@@ -0,0 +1,304 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{collections::HashMap, net::SocketAddr, time::Duration};
+
+use crate::{NetworkError, Node, Payload, Peer};
+use futures::{pin_mut, select, FutureExt};
+use rand::prelude::SliceRandom;
+use snarkvm_algorithms::crh::double_sha256;
+use snarkvm_objects::{BlockHeader, BlockHeaderHash, Storage};
+use tokio::{sync::mpsc, time::Instant};
+
+pub enum SyncInbound {
+    BlockHashes(SocketAddr, Vec<BlockHeaderHash>),
+    Block(SocketAddr, Vec<u8>),
+}
+
+pub struct SyncMaster<S: Storage + Send + Sync + 'static> {
+    node: Node<S>,
+    incoming: mpsc::Receiver<SyncInbound>,
+}
+
+struct SyncBlock {
+    address: SocketAddr,
+    block: Vec<u8>,
+}
+
+impl<S: Storage + Send + Sync + 'static> SyncMaster<S> {
+    pub fn new(node: Node<S>) -> (Self, mpsc::Sender<SyncInbound>) {
+        let (sender, receiver) = mpsc::channel(256);
+        let new = Self {
+            node,
+            incoming: receiver,
+        };
+        (new, sender)
+    }
+
+    async fn find_sync_nodes(&mut self) -> Vec<Peer> {
+        let our_block_height = self.node.expect_sync().current_block_height();
+        let mut interesting_peers = vec![];
+        for node in self.node.peer_book.connected_peers_snapshot().await {
+            if !node.judge() && node.quality.block_height > our_block_height + 1 {
+                interesting_peers.push(node);
+            }
+        }
+        interesting_peers.sort_by(|x, y| y.quality.block_height.cmp(&x.quality.block_height));
+
+        // trim nodes close to us if any are > 10 blocks ahead
+        if let Some(i) = interesting_peers
+            .iter()
+            .position(|x| x.quality.block_height <= our_block_height + 10)
+        {
+            interesting_peers.truncate(i + 1);
+        }
+
+        interesting_peers
+    }
+
+    async fn block_locator_hashes(&mut self) -> Vec<BlockHeaderHash> {
+        match self.node.expect_sync().storage().get_block_locator_hashes() {
+            Ok(block_locator_hashes) => block_locator_hashes,
+            _ => {
+                error!("Unable to get block locator hashes from storage");
+                vec![]
+            }
+        }
+    }
+
+    async fn send_sync_messages(&mut self) -> usize {
+        let sync_nodes = self.find_sync_nodes().await;
+
+        info!("requested block information from {} peers", sync_nodes.len());
+        let block_locator_hashes = self.block_locator_hashes().await;
+        let mut sent = 0usize;
```

Comment:
isn't `sent` equal to` future_set.len()`? we could just calculate that after the loop for brevity